File: Android/Environs/Environs/src/main/java/environs/PortalInfo.java
package environs;
/**
 * PortalInfo
 * ------------------------------------------------------------------
 * Copyright (c) Chi-Tai Dang
 *
 * @author	Chi-Tai Dang
 * @version	1.0
 * @remarks
 *
 * This file is part of the Environs framework developed at the
 * Lab for Human Centered Multimedia of the University of Augsburg.
 * http://hcm-lab.de/environs
 *
 * Environ is free software; you can redistribute it and/or modify
 * it under the terms of the Eclipse Public License v1.0.
 * A copy of the license may be obtained at:
 * http://www.eclipse.org/org/documents/epl-v10.html
 * --------------------------------------------------------------------
 */
import android.annotation.SuppressLint;

/** 
 * A PortalInfo object serves as container for portal information.&nbsp;
 * Environs makes use of such objects to get/set portal details.
 * 
 * @author Chi-Tai Dang
 *
 */
public class PortalInfo 
{
	private static final String className = "PortalInfo . . . . . . .";

	public int deviceID;
	public int portalID;
	public int flags;
	
	public int centerX;
	public int centerY;
	public int width;
	public int height;
	public float orientation;

	PortalInstance	portal = null;


	void NotifyObservers(int notification)
	{
		if (Utils.isDebug) Utils.Log ( 6, className, "NotifyObservers" );

		if ( portal != null )
			portal.NotifyObservers ( notification );
	}


	public void SetSize(int width, int height)
	{
		flags = 0;
		this.width = width;
		this.height = height;
		flags |= Environs.PORTAL_INFO_FLAG_SIZE;

		portal.SetPortalInfo(this);
	}

	public void SetOrientation(float angle)
	{
		flags = 0;
		this.orientation = angle;
		flags |= Environs.PORTAL_INFO_FLAG_ANGLE;

		portal.SetPortalInfo(this);
	}

	public void SetLocation(int centerX, int centerY)
	{
		flags = 0;
		this.centerX = centerX;
		this.centerY = centerY;
		flags |= Environs.PORTAL_INFO_FLAG_LOCATION;

		portal.SetPortalInfo(this);
	}

	public void SetLocation(int centerX, int centerY, float angle)
	{
		flags = 0;
		this.centerX = centerX;
		this.centerY = centerY;
		flags |= Environs.PORTAL_INFO_FLAG_LOCATION;
		this.orientation = angle;
		flags |= Environs.PORTAL_INFO_FLAG_ANGLE;

		portal.SetPortalInfo(this);
	}

	public void Set(int centerX, int centerY, float angle, int width, int height)
	{
		flags = 0;
		this.centerX = centerX;
		this.centerY = centerY;
		flags |= Environs.PORTAL_INFO_FLAG_LOCATION;
		this.orientation = angle;
		flags |= Environs.PORTAL_INFO_FLAG_ANGLE;
		this.width = width;
		this.height = height;
		flags |= Environs.PORTAL_INFO_FLAG_SIZE;

		portal.SetPortalInfo(this);
	}

	
	@SuppressLint("DefaultLocale")
	public String toString() {
		
        return ("Portal: center coordinates [ " + centerX + " / " + centerY + " ], size [ " 
        + width + " / " + height + " ], orientation [ " + orientation + " ]");
    }

	public String ToString() {
		return toString();
	}


	boolean Update(int notification, PortalInfo info)
	{
		//if (Utils.isDebug) Utils.Log ( 0, className, "Update: " + info.ToString () );

		boolean changedl = false;
		boolean changeds = false;

		if (portalID != info.portalID && info.portalID > 0)
			portalID = info.portalID;

		if (notification == 0 || notification == Environs.NOTIFY_PORTAL_LOCATION_CHANGED)
		{
			if (info.centerX != centerX)
			{
				changedl = true;
				centerX = info.centerX;
			}

			if (info.centerY != centerY)
			{
				changedl = true;
				centerY = info.centerY;
			}

			if (info.orientation != orientation)
			{
				changedl = true;
				orientation = info.orientation;
			}

			if (changedl)
				NotifyObservers(Environs.NOTIFY_PORTAL_LOCATION_CHANGED);
		}

		if (notification == 0 || notification == Environs.NOTIFY_PORTAL_SIZE_CHANGED)
		{
			if (info.width != width)
			{
				changeds = true;
				width = info.width;
			}

			if (info.height != height)
			{
				changeds = true;
				height = info.height;
			}

			if (changeds)
				NotifyObservers(Environs.NOTIFY_PORTAL_SIZE_CHANGED);
		}
		return (changedl | changeds);
	}
}

File: Native/Portal/Portal.Info.cpp
/**
 * A PortalInfo object serves as container for portal information.
 * Environs makes use of such objects to get/set portal details.
 * ------------------------------------------------------------------
 * Copyright (c) Chi-Tai Dang
 *
 * @author	Chi-Tai Dang
 * @version	1.0
 * @remarks
 *
 * This file is part of the Environs framework developed at the
 * Lab for Human Centered Multimedia of the University of Augsburg.
 * http://hcm-lab.de/environs
 *
 * Environ is free software; you can redistribute it and/or modify
 * it under the terms of the Eclipse Public License v1.0.
 * A copy of the license may be obtained at:
 * http://www.eclipse.org/org/documents/epl-v10.html
 * --------------------------------------------------------------------
 */
#include "stdafx.h"

/// Compiler flag that enables verbose debug output
#ifndef NDEBUG
//#   define DEBUGVERB
//#   define DEBUGVERBVerb
#endif

#ifndef CLI_CPP
#	include "Interop.h"
#	include "Interop/Smart.Pointer.h"
#	include "Environs.Platforms.h"

#	include "Environs.Obj.h"
#endif

#	ifdef CLI_CPP
#		include "Environs.Cli.Forwards.h"
#		include "Environs.Native.h"
#		include "Environs.Utils.h"

#		include "Environs.Cli.h"
#		include "Environs.Lib.h"
#		include "Portal.Instance.Cli.h"
#	else
#		include "Environs.Observer.h"
#		include "Portal.Instance.h"

		using namespace std;
#	endif


// The TAG for prepending to log messages
#define CLASS_NAME	"Portal.Info. . . . . . ."


namespace environs 
{
	PortalInfo::PortalInfo ( )
	{
		CVerb ( "Construct" );
        
		C_Only ( Zero ( base ); )
       
        hEnvirons	= 0;
		portal		= nill;
    }
    
    
    void PortalInfo::NotifyObservers ( environs::Notify::Portale_t notification )
    {
        CVerb ( "NotifyObservers" );

		CLIBSPACE PortalInstance OBJ_ptr portalInstance = portal;

		if ( portalInstance != nill )
			portalInstance->NotifyObservers ( notification );
    }
    
    
    void PortalInfo::SetSize ( int width, int height )
    {
        base.flags = 0;
        base.width = width;
        base.height = height;
        base.flags |= PORTAL_INFO_FLAG_SIZE;

		CLIBSPACE PortalInstance::SetPortalInfo ( hEnvirons, Addr_of ( base ) );
    }
    

    void PortalInfo::SetOrientation ( float angle )
    {
        base.flags = 0;
        base.orientation = angle;
        base.flags |= PORTAL_INFO_FLAG_ANGLE;

		CLIBSPACE PortalInstance::SetPortalInfo ( hEnvirons, Addr_of ( base ) );
    }
    

    void PortalInfo::SetLocation ( int centerX, int centerY )
    {
        base.flags = 0;
        base.centerX = centerX;
        base.centerY = centerY;
		base.flags |= PORTAL_INFO_FLAG_LOCATION;

		CLIBSPACE PortalInstance::SetPortalInfo ( hEnvirons, Addr_of ( base ) );
    }
    

    void PortalInfo::SetLocation ( int centerX, int centerY, float angle )
    {
        base.flags = 0;
        base.centerX = centerX;
        base.centerY = centerY;
        base.orientation = angle;
		base.flags = PORTAL_INFO_FLAG_ANGLE | PORTAL_INFO_FLAG_LOCATION;

		CLIBSPACE PortalInstance::SetPortalInfo ( hEnvirons, Addr_of ( base ) );
    }
    

    void PortalInfo::Set ( int centerX, int centerY, float angle, int width, int height )
    {
        base.flags = 0;
        base.centerX = centerX;
        base.centerY = centerY;
        base.orientation = angle;
        base.width = width;
        base.height = height;
		base.flags = PORTAL_INFO_FLAG_LOCATION | PORTAL_INFO_FLAG_ANGLE | PORTAL_INFO_FLAG_SIZE;

		CLIBSPACE PortalInstance::SetPortalInfo ( hEnvirons, Addr_of ( base ) );
    }
    
    
	STRING_T PortalInfo::toString ()
    {
#ifdef CLI_CPP
		StringBuilder ^ sbl = gcnew StringBuilder ( "Portal: center coordinates [ " );

		sbl->Append ( base.centerX )->Append ( " / " )->Append ( base.centerY )
			->Append ( " ], size [ " )->Append ( base.width )->Append ( " / " )
			->Append ( base.height )->Append ( " ], orientation [ " )->Append ( base.orientation )->Append ( " ]" );

		return sbl->ToString ();
#else            
		char buffer [ 256 ];

		if ( snprintf ( buffer, 256, "Portal: center coordinates [ %i / %i ], size [ %i / %i ], orientation [ %f ]",
			base.centerX, base.centerY, base.width, base.height, base.orientation ) < 0 )
			*buffer = 0;

		return string ( buffer );
#endif
    }


#ifdef CLI_CPP
	STRING_T PortalInfo::ToString ()
	{
		return toString ();
	}
#endif
    
    
    bool PortalInfo::Update ( environs::Notify::Portale_t notification, PortalInfoBase OBJ_ptr src )
    {
        bool changedl = false;
        bool changeds = false;

        if (base.portalID != src->portalID && src->portalID > 0)
            base.portalID = src->portalID;
        
        if (notification == Notify::Portale::Zero || notification == Notify::Portale::LocationChanged )
        {
            if (base.centerX != src->centerX)
            {
                changedl = true;
                base.centerX = src->centerX;
            }
            
            if (base.centerY != src->centerY)
            {
                changedl = true;
                base.centerY = src->centerY;
            }
            
            if (base.orientation != src->orientation)
            {
                changedl = true;
                base.orientation= src->orientation;
            }
            
            if (changedl)
                NotifyObservers ( Notify::Portale::LocationChanged );
        }
        
        if (notification == Notify::Portale::Zero || notification == Notify::Portale::SizeChanged )
        {
            if (base.width != src->width)
            {
                changeds = true;
                base.width= src->width;
            }
            
            if (base.height != src->height)
            {
                changeds = true;
                base.height= src->height;
            }
            
            if (changeds)
                NotifyObservers ( Notify::Portale::SizeChanged );
        }
        return (changedl | changeds);
    }
    
} /// -> namespace environs



